File: backend/app/schemas/projection.py
"""
Pydantic schemas for the Projection Computation API.

Defines the request/response contracts. The frontend sends a ProjectionRequest
and receives a ProjectionResponse containing pre-computed render instructions.
The frontend has zero knowledge of projection geometry — it just draws.
"""

from __future__ import annotations

from enum import Enum
from typing import Literal

from pydantic import BaseModel, Field


# ============================================================
# Enums
# ============================================================

class SolidType(str, Enum):
    """Supported solid types — maps to getSidesCount() in core.js:550-556."""
    TRIANGULAR_PRISM = "triangular-prism"
    SQUARE_PRISM = "square-prism"
    PENTAGONAL_PRISM = "pentagonal-prism"
    HEXAGONAL_PRISM = "hexagonal-prism"
    TRIANGULAR_PYRAMID = "triangular-pyramid"
    SQUARE_PYRAMID = "square-pyramid"
    PENTAGONAL_PYRAMID = "pentagonal-pyramid"
    HEXAGONAL_PYRAMID = "hexagonal-pyramid"


class CaseType(str, Enum):
    """Projection case types."""
    A = "A"  # Axis perpendicular to HP
    B = "B"  # Axis perpendicular to VP
    C = "C"  # Axis inclined to HP
    D = "D"  # Axis inclined to both HP and VP


class RestingOn(str, Enum):
    """Resting condition for Case C/D — from addRestingInputCaseC() core.js:190-207."""
    BASE_EDGE = "base-edge"
    BASE_CORNER = "base-corner"


# ============================================================
# Request
# ============================================================

class ProjectionRequest(BaseModel):
    """
    Input parameters for projection computation.

    Maps directly to the state object in core.js:13-33 and the dynamic
    inputs created by updateDynamicInputs() in core.js:135-168.
    """
    solid_type: SolidType = Field(
        ...,
        description="Type of solid (e.g., 'hexagonal-prism')",
    )
    case_type: CaseType = Field(
        ...,
        description="Projection case (A, B, C, or D)",
    )
    base_edge: float = Field(
        default=40.0,
        gt=0,
        le=200,
        description="Length of one base edge in drawing units",
    )
    axis_length: float = Field(
        default=80.0,
        gt=0,
        le=400,
        description="Length of solid axis in drawing units",
    )
    edge_angle: float = Field(
        default=30.0,
        ge=0,
        le=90,
        description="Edge angle with VP in degrees (Case A)",
    )
    axis_angle_hp: float = Field(
        default=45.0,
        ge=0,
        le=90,
        description="Axis inclination with HP in degrees (Case C/D)",
    )
    axis_angle_vp: float = Field(
        default=0.0,
        ge=0,
        le=90,
        description="Axis inclination with VP in degrees (Case D)",
    )
    resting_on: RestingOn = Field(
        default=RestingOn.BASE_EDGE,
        description="Resting condition for Case C/D",
    )
    canvas_width: float = Field(
        default=1200.0,
        gt=0,
        description="Available canvas width in pixels",
    )
    canvas_height: float = Field(
        default=700.0,
        gt=0,
        description="Available canvas height in pixels",
    )


# ============================================================
# Render Elements (discriminated union by 'type' field)
# ============================================================

class LineElement(BaseModel):
    """A line segment — maps to drawLine() in core.js:437-460."""
    type: Literal["line"] = "line"
    x1: float
    y1: float
    x2: float
    y2: float
    style: Literal["visible", "hidden", "construction"] = "visible"


class PolygonElement(BaseModel):
    """A closed/open polygon — maps to polygon drawing in caseA.js:122-133."""
    type: Literal["polygon"] = "polygon"
    points: list[dict[str, float]]  # [{"x": ..., "y": ...}, ...]
    style: Literal["visible", "hidden", "construction"] = "visible"
    closed: bool = True


class PointElement(BaseModel):
    """A labeled point — maps to drawPoint() in core.js:462-475."""
    type: Literal["point"] = "point"
    x: float
    y: float
    label: str = ""
    radius: float = 2.0


class LabelElement(BaseModel):
    """A text label — for annotations not attached to a point."""
    type: Literal["label"] = "label"
    x: float
    y: float
    text: str
    font_size: float = 12.0


class ArcElement(BaseModel):
    """An arc — maps to drawAngleArc() in core.js:499-510."""
    type: Literal["arc"] = "arc"
    center_x: float
    center_y: float
    radius: float
    start_angle: float  # in degrees
    end_angle: float  # in degrees


class ArrowElement(BaseModel):
    """An arrow head — maps to drawArrow() in core.js:423-435."""
    type: Literal["arrow"] = "arrow"
    from_x: float
    from_y: float
    to_x: float
    to_y: float


# Discriminated union type
RenderElement = LineElement | PolygonElement | PointElement | LabelElement | ArcElement | ArrowElement


# ============================================================
# Response
# ============================================================

class StepInstruction(BaseModel):
    """
    One step in the projection sequence.

    Each step contains all elements needed to render that step on canvas.
    Steps are cumulative — step N includes all drawing from steps 1..N.
    This matches the behavior of drawStep() in core.js:365-383.
    """
    step_number: int = Field(..., ge=1)
    title: str
    description: str
    elements: list[
        LineElement | PolygonElement | PointElement | LabelElement | ArcElement | ArrowElement
    ]


class SolidProperties(BaseModel):
    """Computed properties of the solid for frontend display."""
    sides: int
    is_prism: bool
    is_pyramid: bool
    circumradius: float | None = None


class ProjectionMetadata(BaseModel):
    """Metadata about the computation result."""
    computed_beta: float | None = None
    computed_xy_length: float
    solid_properties: SolidProperties


class ProjectionResponse(BaseModel):
    """
    Complete projection computation result.

    Contains pre-computed render instructions for each step.
    The frontend iterates over steps and renders elements via Canvas 2D API.
    """
    total_steps: int = Field(..., ge=0)
    steps: list[StepInstruction]
    metadata: ProjectionMetadata

File: frontend/src/types/projection.ts
/**
 * TypeScript types for the Projection Computation API.
 *
 * Exact mirror of backend Pydantic schemas in:
 *   backend/app/schemas/projection.py (lines 1-220)
 *
 * Every field name, type, and literal value matches the Python model.
 * The frontend has zero knowledge of projection geometry — it only
 * interprets these render instructions.
 */

// ============================================================
// Enums (as string union types)
// ============================================================

/** Maps to SolidType enum — projection.py:21-30 */
export type SolidType =
    | 'triangular-prism'
    | 'square-prism'
    | 'pentagonal-prism'
    | 'hexagonal-prism'
    | 'triangular-pyramid'
    | 'square-pyramid'
    | 'pentagonal-pyramid'
    | 'hexagonal-pyramid';

/** Maps to CaseType enum — projection.py:33-38 */
export type CaseType = 'A' | 'B' | 'C' | 'D';

/** Maps to RestingOn enum — projection.py:41-44 */
export type RestingOn = 'base-edge' | 'base-corner';

/** Style applied to lines and polygons — projection.py:123, 130 */
export type ElementStyle = 'visible' | 'hidden' | 'construction';

// ============================================================
// Render Element Types (discriminated union by 'type' field)
// ============================================================

/** Maps to LineElement — projection.py:116-123 */
export interface LineElement {
    type: 'line';
    x1: number;
    y1: number;
    x2: number;
    y2: number;
    style: ElementStyle;
}

/** Maps to PolygonElement — projection.py:126-131 */
export interface PolygonElement {
    type: 'polygon';
    points: Array<{ x: number; y: number }>;
    style: ElementStyle;
    closed: boolean;
}

/** Maps to PointElement — projection.py:134-140 */
export interface PointElement {
    type: 'point';
    x: number;
    y: number;
    label: string;
    radius: number;
}

/** Maps to LabelElement — projection.py:143-149 */
export interface LabelElement {
    type: 'label';
    x: number;
    y: number;
    text: string;
    font_size: number;
}

/** Maps to ArcElement — projection.py:152-159 */
export interface ArcElement {
    type: 'arc';
    center_x: number;
    center_y: number;
    radius: number;
    start_angle: number;
    end_angle: number;
}

/** Maps to ArrowElement — projection.py:162-168 */
export interface ArrowElement {
    type: 'arrow';
    from_x: number;
    from_y: number;
    to_x: number;
    to_y: number;
}

/** Discriminated union — projection.py:172 */
export type RenderElement =
    | LineElement
    | PolygonElement
    | PointElement
    | LabelElement
    | ArcElement
    | ArrowElement;

// ============================================================
// Response Types
// ============================================================

/** Maps to StepInstruction — projection.py:179-192 */
export interface StepInstruction {
    step_number: number;
    title: string;
    description: string;
    elements: RenderElement[];
}

/** Maps to SolidProperties — projection.py:195-200 */
export interface SolidProperties {
    sides: number;
    is_prism: boolean;
    is_pyramid: boolean;
    circumradius: number | null;
}

/** Maps to ProjectionMetadata — projection.py:203-207 */
export interface ProjectionMetadata {
    computed_beta: number | null;
    computed_xy_length: number;
    solid_properties: SolidProperties;
}

/** Maps to ProjectionResponse — projection.py:210-219 */
export interface ProjectionResponse {
    total_steps: number;
    steps: StepInstruction[];
    metadata: ProjectionMetadata;
}

// ============================================================
// Request Type
// ============================================================

/** Maps to ProjectionRequest — projection.py:51-109 */
export interface ProjectionRequest {
    solid_type: SolidType;
    case_type: CaseType;
    base_edge: number;
    axis_length: number;
    edge_angle: number;
    axis_angle_hp: number;
    axis_angle_vp: number;
    resting_on: RestingOn;
    canvas_width: number;
    canvas_height: number;
}

// ============================================================
// UI Configuration Constants
// ============================================================

/** Solid type options for the select dropdown — matches HTML lines 88-100 */
export const SOLID_OPTIONS: Array<{ value: SolidType; label: string }> = [
    { value: 'triangular-prism', label: 'Triangular Prism' },
    { value: 'square-prism', label: 'Square Prism' },
    { value: 'pentagonal-prism', label: 'Pentagonal Prism' },
    { value: 'hexagonal-prism', label: 'Hexagonal Prism' },
    { value: 'triangular-pyramid', label: 'Triangular Pyramid' },
    { value: 'square-pyramid', label: 'Square Pyramid' },
    { value: 'pentagonal-pyramid', label: 'Pentagonal Pyramid' },
    { value: 'hexagonal-pyramid', label: 'Hexagonal Pyramid' },
];

/** Case type options — matches HTML lines 106-112 */
export const CASE_OPTIONS: Array<{ value: CaseType; label: string }> = [
    { value: 'A', label: 'Case A — Axis ⊥ HP' },
    { value: 'B', label: 'Case B — Axis ⊥ VP' },
    { value: 'C', label: 'Case C — Axis Inclined to HP' },
    { value: 'D', label: 'Case D — Axis Inclined to VP' },
];

/** Resting condition options */
export const RESTING_OPTIONS: Array<{ value: RestingOn; label: string }> = [
    { value: 'base-edge', label: 'Resting on Base Edge' },
    { value: 'base-corner', label: 'Resting on Base Corner' },
];
